cmd/calico-ipam-upgrade/main.go

```diff
@@ -0,0 +1,333 @@
+			if err = c.IPAM().AssignIP(ctxt, ipam.AssignIPArgs{
+				IP:       *ipipTunnelAddr,
+				Hostname: nodename,
+				Attrs: map[string]string{
+					"node":           nodename,
+					"ipipTunnelAddr": "true",
+				},
+			}); err != nil {
+				if _, ok := err.(errors.ErrorResourceAlreadyExists); !ok {
+					return fmt.Errorf("failed to get add IPIP tunnel addr %s: %s", node.Spec.BGP.IPv4IPIPTunnelAddr, err)
+				}
+				log.Info("IPIP tunnel address already assigned, continuing...")
+			}
+
+			// Save the calico node object to the datastore.
+			if node, err = c.Nodes().Update(ctxt, node, options.SetOptions{}); err != nil {
+				return fmt.Errorf("failed to save newly updated node object: %s", err)
+			}
+
+			log.Info("Successfully assigned and updated IPIP tunnel address")
+		}
+	}
+
+	// Open k8s-pod-directory to check for emptiness.
+	ipamDir, err := os.Open(ipAllocPath)
+	if err != nil {
+		return fmt.Errorf("failed to open k8s-pod-network dir: %s", err)
+	}
+
+	// Ensure that the k8s-pod-directory file pointer is closed
+	// and also deleted if this migration ran successfully.
+	defer func() {
+		if closeErr := ipamDir.Close(); closeErr != nil {
+			log.WithError(err).Errorf("failed to close k8s-pod-network directory")
+		}
+		if err == nil {
+			if err = os.Remove(ipAllocPath); err != nil {
+				log.WithError(err).Error("failed to remove k8s-pod-network directory")
+			}
+		}
+	}()
+
+	// Check if the directory is empty.
+	if _, err = ipamDir.Readdirnames(1); err != nil {
+		if err == io.EOF {
+			log.Info("k8s-pod-network dir empty; no migration necessary, removing directory and exiting...")
+			return nil
+		}
+		return fmt.Errorf("failed to read k8s-pod-network dir names: %s", err)
+	}
+
+	// Disable cni by setting datastoreReady to false.
+	clusterInfo, err := c.ClusterInformation().Get(ctxt, "default", options.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to fetch cluster information: %s", err)
+	}
+	if *clusterInfo.Spec.DatastoreReady {
+		*clusterInfo.Spec.DatastoreReady = false
+		for i := uint(0); i < 5; i++ {
+			if clusterInfo, err = c.ClusterInformation().Update(ctxt, clusterInfo, options.SetOptions{}); err != nil {
+				if _, ok := err.(errors.ErrorResourceUpdateConflict); ok {
+					log.Info("Encountered update conflict, retrying...")
+					time.Sleep((1 << i) * time.Second)
+				}
+				return fmt.Errorf("failed to disable cluster: %s", err)
+			}
+			break
+		}
+	}
+
+	// Also disable cni by disabling the binaries.
+	for _, binary := range binariesToDisable {
+		dir, file := path.Split(binary)
+		disabledBinary := path.Join(dir, file+".disabled")
+		if err = os.Rename(binary, disabledBinary); err != nil {
+			if os.IsNotExist(err) {
+				log.WithField("binary", file).Info("binary not found, probably already disabled, continuing...")
+				continue
+			}
+			return fmt.Errorf("failed to disable binary %s: %s", binary, err)
+		}
+	}
+	time.Sleep(3 * time.Second)
```

Comment:
I think we should up this to 5 seconds to be safe. It's somewhat arbitrary ("magic number") but 3 feels like it might not be quite enough.